### ISSUE TITLE
[FEATURE] metrics on subsets of X,y 

### ISSUE THREAD

**Issue by koaning** [koaning/scikit-lego#209]
There may be a critical subset of your data for which you want to make sure that your model works well. For example; you might want to have a metric that allows you to measure the accuracy for your model for both males and females. 

The goal of this issue thread is to discuss if we might discover a nice solution for this issue. Potentially via a meta metric but other options are also possible.

**Comment by janvdvegt** [comment]
As mentioned yesterday, I'm interested in looking into this feature. Question: this meta metric you talk about means those closures that return the actual metric method?

My knowledge of the scikit internals is not deep enough yet, so a question with regards to metrics in general. Is it possible to return a set of metrics or does every metric represent one scalar? In case of the latter, there is a user experience overhead where they would need to create a metric for every categorical value they are interested in. If this first option is available, they could specify some sort of method that groups the data into multiple subsets that we could apply the same metric over. It looks like returning a list of metrics could be a viable strategy but I'm wondering if that is clear enough to users. What do you think?

In either case, my proposal would be to have a `subset_metric` method that the user passes the following arguments to:

- `metric` which is the metric (or maybe also a list of metrics if we can do multiples at once) to apply to every subset
- `name` which is an optional name for the metric so that you could call it something like `sliced_accuracy` instead of `accuracy` to indicate that it is only a subset

If we can create multiple slices at the same time:

- `subset` which is either a string that corresponds to a categorical column or a method that takes the whole (X, y) space and returns a numpy array with shape (X.shape[0]) of strings / categories which are then grouped. For each of these groups the metric will be calculated and assigned to the metric `{metric_name}_{group_name}` in case of no specified name

Otherwise:

- `subset` would need to be some method that returns True / False per sample based on (X, y). In case of Pandas or named numpy arrays this could be a string and we could have an additional argument of the value that it would need to equate with

This method would then return a (list of) metric(s) which conform the scikit-learn API guidelines.

What do you think of this approach?

**Comment by koaning** [comment]
The idea that you're proposing has merit to it. I think you don't need to pass the name to the generating function though. The gridsearch takes care of that.

It may be clearer if we start this discussion with an example of how a person might use a `subset_metric`. 

```python
from sklearn.metrics import accuracy_score
from sklego.metrics import subset_metric

subset_accuracy_score = subset_metric(subset_picker=lambda d: d['income'] == 'low', accuracy_score)

fair_classifier = GridSearchCV(estimator=FairClassifier(sensitive_cols="lstat", covariance_threshold=0.5),
                               param_grid={"estimator__covariance_threshold":
                                           np.linspace(0.01, 1.00, 100)},
                               cv=5,
                               refit="accuracy_score",
                               return_train_score=True,
                               scoring={"p_percent_score": p_percent_score('lstat'),
                                        "accuracy_score": make_scorer(accuracy_score),
                                        "subset_accuracy_score": make_scorer(subset_accuracy_score)})
```

**Comment by MBrouns** [comment]
> Question: this meta metric you talk about means those closures that return the actual metric method? 

yes!

>  Is it possible to return a set of metrics or does every metric represent one scalar?

As far as I know it's only possible to return a single scalar. The gridsearch takes a dict though so if the meta-metric returns a dict you can use `{**subsetmetric(...), my_other_metric}`

> name which is an optional name for the metric so that you could call it something like sliced_accuracy instead of accuracy to indicate that it is only a subset\

Gridsearch handles this itself quite nicely

**Comment by janvdvegt** [comment]
> Note that the intention here is to have a discussion and to clarify. Feel free to point it out if you feel like I misinterpreted your approach or if you think this approach has problems to it.

Noted, my idea about (the possibility of) returning multiple scores is that you are usually not interested in just one subset, it's usually multiple slices over the same dimension from my perspective. We are not only interested in accuracy of women or people of color, we want to compare those metrics to their counter parts. That's where my suggestion of returning multiple categories with the lambda is coming from. The metrics corresponding to each subset would then be named after each subset. An example given the dictionary return from Matthijs would then be:

`def age_category(age: pd.Series):
    return np.where(age < 18, 'child', np.where(age < 65, 'adult', 'senior)

metric_dict = subset_metric(subset_picker=age_category, accuracy_score)`

This metric_dict would then contain the following key - value pairs:

`{'accuracy_score_child': accuracy_score_child,
'accuracy_score_adult': accuracy_score_adult,
'accuracy_score_senior': accuracy_score_senior}`

The only thing in my approach would be how to approach subsets we are not actually interested in. My knowledge about numpy dtypes is not good enough to know exactly what would be a good representation but returning `np.nan` for those cases could be a clean approach, where we take the non-nan unique values as all the metrics we should generate.

Maybe my hypothesis is wrong that in most cases you would be interested in different slices over the same dimension, what do you think?

**Comment by koaning** [comment]
> The metrics corresponding to each subset would then be named after each subset. 

I mean ... technically ... you could still do that with my previous approach by using a one-liner. 

```python
{'accuracy_subset_' + k: subset_metric(picker=lambda d: d['type'] == k, accuracy_score) 
  for k in ['child', 'adult', 'senior']}
```

> The only thing in my approach would be how to approach subsets we are not actually interested in.

This seems to also be fixed with the dict comprehension. 

### But what about multiple columns? 

You could repeat this approach for other columns. 

```python
{'accuracy_subset_' + k: subset_metric(picker=lambda d: d['age'] == k, accuracy_score) 
  for k in ['child', 'adult', 'senior']}

{'accuracy_subset_' + k: subset_metric(picker=lambda d: d['gender'] == k, accuracy_score) 
  for k in ['m', 'f', 'x']}
```

**Comment by janvdvegt** [comment]
Yeah that is a clean approach too, my reasoning behind it was that for multiple categories you would recalculate the mask even though very commonly you would want multiple slices anyway. But I suggest we implement the binary mask approach first and then see if it makes sense to add the multiple slices as an additional feature.

I'll setup my environment and see if I can get the tests running. After that I will see if I can make a quick initial implementation.

**Comment by janvdvegt** [comment]
I added the feature and wrote a number of tests. Two questions:

1) What do I need to be able to do to push my changes to a branch and create a pull request

2) What would be preferred with regards to documentation?

**Comment by koaning** [comment]
@janvdvegt did you fork? you should be able to create a PR from github. please ensure that the unit tests run. 

if you have documentation ready we wouldn't mind having a look. another option is to first get this merged and worry about documentation later.

**Comment by janvdvegt** [comment]
I did not fork, I cloned and created a branch. I'll check how to turn it into a fork and then I'll create a PR that way. All the unit tests are running. I added a docstring to the new method, other documentation can come later then.

**Comment by koaning** [comment]
Yeah that's fine. We sometimes do soft releases that we want to test "out there". Only the items that are in a getting started guide are considered to be "formal" in that sense.
